inline-c-cpp/src/Language/C/Inline/Cpp/Exceptions.hs

```diff
@@ -0,0 +1,68 @@
+-- | A module that contains exception-safe equivalents of @inline-c@ QuasiQuoters.
+
+{-# LANGUAGE TemplateHaskell #-}
+
+module Language.C.Inline.Cpp.Exceptions
+  ( ForeignException(..)
+  , catchBlock
+  ) where
+
+import           Control.Exception.Safe
+import           Control.Monad
+import qualified Language.C.Inline as C
+import           Language.Haskell.TH
+import           Language.Haskell.TH.Quote
+import           Foreign
+import           Foreign.C
+
+-- | An exception thrown in C++ code, with the message from @what()@.
+newtype ForeignException = ForeignException String
```

Comment:
would it be possible to distinguish between `std::exception`s and non-exceptions here? e.g. with a sum type in `CppException`.